test_util/azure.py

```diff
@@ -35,77 +35,123 @@ def validate_hostname_prefix(prefix):
+            # Ensure the resource group will be deleted if the following steps fail
+            stack.callback(self.rmc.resource_groups.delete, group_name)
             log.info('Resource group created: {}'.format(group_name))
-
-            template_params = {
-                'sshRSAPublicKey': public_key,
-                # must be lower or validation will fail
-                'masterEndpointDNSNamePrefix': master_prefix,
-                'agentEndpointDNSNamePrefix': agent_prefix,
-                'linuxAdminUsername': admin_name,
-                'agentVMSize': vm_size,
-                'agentCount': agent_count,
-                'nameSuffix': name_suffix,
-                'oauthEnabled': repr(oauth_enabled).lower(),
-                # oauth uses string, vm diagnostic uses bool
-                'enableVMDiagnostics': vm_diagnostics_enabled}
-            log.info('Provided template parameters: {}'.format(template_params))
-            # azure requires that parameters be provided as {key: {'value': value}}
-            template_parameters = {k: {'value': v} for k, v in template_params.items()}
-            deployment_properties = DeploymentProperties(
-                template_link=TemplateLink(uri=template_uri),
-                mode=DeploymentMode.incremental,
-                parameters=template_parameters)
+            deployment_properties = self.create_deployment_properties(
+                template_url, parameters)
             log.info('Checking with Azure to validate template deployment')
-            result = azure_wrapper.rmc.deployments.validate(
+            result = self.rmc.deployments.validate(
                 group_name, deployment_name, properties=deployment_properties)
             if result.error:
-                for details in result.error.details:
-                    log.error('{}: {}'.format(details.code, details.message))
-                error_msgs = '\n'.join(['{}: {}'.format(d.code, d.message) for d in result.error.details])
-                raise Exception("Template verification failed!\n{}".format(error_msgs))
+                log.critical('{}: {}'.format(result.error.code, result.error.message))
+                raise Exception("Template verification failed!\n{}".format(result.error))
             log.info('Template successfully validated')
             log.info('Starting template deployment')
-            azure_wrapper.rmc.deployments.create_or_update(
+            self.rmc.deployments.create_or_update(
                 group_name, deployment_name, deployment_properties)
             stack.pop_all()
+        log.info('Successfully started template deployment')
+
+    def create_deployment_properties(self, template_url, parameters):
+        """ Pulls the targeted template, checks parameter specs and casts
+        user provided parameters to the appropriate type. Assertion is raised
+        if there are unused parameters
+        """
+        user_parameters = copy.copy(parameters)
```

Comment:
If there are going to be list of lists or dictionary with list, it should be `copy.deepcopy`

1. My rule of thumb. use deepcopy unless you worry about space explosion.